File: code/postgresql/monitoring/log_analyzer.py
#!/usr/bin/env python3
"""
PostgreSQL ログ分析ツール

使用方法:
    python log_analyzer.py /var/log/postgresql/postgresql.log
"""

import re
import csv
from datetime import datetime
from collections import defaultdict, Counter
import argparse

class PostgreSQLLogAnalyzer:
    def __init__(self, log_file):
        self.log_file = log_file
        self.errors = []
        self.slow_queries = []
        self.connections = defaultdict(int)
        self.query_patterns = Counter()
        
    def parse_log(self):
        """ログファイルを解析"""
        error_pattern = re.compile(r'ERROR:|FATAL:|PANIC:')
        slow_query_pattern = re.compile(r'duration: (\d+\.\d+) ms')
        connection_pattern = re.compile(r'connection (authorized|received)')
        
        with open(self.log_file, 'r', encoding='utf-8', errors='ignore') as f:
            for line in f:
                # エラー検出
                if error_pattern.search(line):
                    self.errors.append(line.strip())
                
                # スロークエリ検出
                slow_match = slow_query_pattern.search(line)
                if slow_match:
                    duration = float(slow_match.group(1))
                    if duration > 1000:  # 1秒以上
                        self.slow_queries.append({
                            'duration': duration,
                            'line': line.strip()
                        })
                
                # 接続パターン分析
                if connection_pattern.search(line):
                    user_match = re.search(r'user=(\w+)', line)
                    if user_match:
                        self.connections[user_match.group(1)] += 1
    
    def analyze_errors(self):
        """エラー分析"""
        print("=" * 80)
        print("エラー分析")
        print("=" * 80)
        
        error_types = Counter()
        for error in self.errors:
            if 'deadlock' in error.lower():
                error_types['Deadlock'] += 1
            elif 'connection' in error.lower():
                error_types['Connection Error'] += 1
            elif 'syntax' in error.lower():
                error_types['Syntax Error'] += 1
            elif 'permission' in error.lower():
                error_types['Permission Denied'] += 1
            else:
                error_types['Other'] += 1
        
        print(f"\n総エラー数: {len(self.errors)}")
        print("\nエラータイプ別集計:")
        for error_type, count in error_types.most_common():
            print(f"  {error_type}: {count}")
        
        if self.errors:
            print("\n最新のエラー（最大5件）:")
            for error in self.errors[-5:]:
                print(f"  {error[:200]}")
    
    def analyze_slow_queries(self):
        """スロークエリ分析"""
        print("\n" + "=" * 80)
        print("スロークエリ分析")
        print("=" * 80)
        
        if not self.slow_queries:
            print("\nスロークエリは検出されませんでした。")
            return
        
        print(f"\nスロークエリ数: {len(self.slow_queries)}")
        
        sorted_queries = sorted(self.slow_queries, key=lambda x: x['duration'], reverse=True)
        
        print("\n最も遅いクエリ（TOP 5）:")
        for i, query in enumerate(sorted_queries[:5], 1):
            print(f"\n{i}. 実行時間: {query['duration']:.2f} ms")
            print(f"   {query['line'][:300]}")
    
    def analyze_connections(self):
        """接続パターン分析"""
        print("\n" + "=" * 80)
        print("接続パターン分析")
        print("=" * 80)
        
        print("\nユーザー別接続数:")
        for user, count in sorted(self.connections.items(), key=lambda x: x[1], reverse=True):
            print(f"  {user}: {count}")
    
    def generate_report(self):
        """総合レポート生成"""
        print("\n" + "=" * 80)
        print("PostgreSQL ログ分析レポート")
        print(f"分析ファイル: {self.log_file}")
        print(f"分析日時: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
        print("=" * 80)
        
        self.parse_log()
        self.analyze_errors()
        self.analyze_slow_queries()
        self.analyze_connections()
        
        # 推奨アクション
        print("\n" + "=" * 80)
        print("推奨アクション")
        print("=" * 80)
        
        if len(self.errors) > 100:
            print("⚠️  エラーが多発しています。アプリケーションまたはDB設定を確認してください。")
        
        if len(self.slow_queries) > 50:
            print("⚠️  スロークエリが多数検出されました。インデックス最適化を検討してください。")
        
        if max(self.connections.values(), default=0) > 1000:
            print("⚠️  特定ユーザーの接続数が多い。コネクションプーリングを確認してください。")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='PostgreSQL ログ分析ツール')
    parser.add_argument('log_file', help='分析対象のログファイル')
    args = parser.parse_args()
    
    analyzer = PostgreSQLLogAnalyzer(args.log_file)
    analyzer.generate_report()

File: code/postgresql/monitoring/log_analyzer.go
package main

/*
PostgreSQL ログ分析ツール

使用方法:
    go run log_analyzer.go /var/log/postgresql/postgresql.log
*/

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"
)

type SlowQuery struct {
	Duration float64
	Line     string
}

type PostgreSQLLogAnalyzer struct {
	LogFile     string
	Errors      []string
	SlowQueries []SlowQuery
	Connections map[string]int
}

func NewPostgreSQLLogAnalyzer(logFile string) *PostgreSQLLogAnalyzer {
	return &PostgreSQLLogAnalyzer{
		LogFile:     logFile,
		Errors:      []string{},
		SlowQueries: []SlowQuery{},
		Connections: make(map[string]int),
	}
}

func (a *PostgreSQLLogAnalyzer) ParseLog() error {
	file, err := os.Open(a.LogFile)
	if err != nil {
		return err
	}
	defer file.Close()

	errorPattern := regexp.MustCompile(`ERROR:|FATAL:|PANIC:`)
	slowQueryPattern := regexp.MustCompile(`duration: (\d+\.?\d*) ms`)
	connectionPattern := regexp.MustCompile(`connection (authorized|received)`)
	userPattern := regexp.MustCompile(`user=(\w+)`)

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()

		// エラー検出
		if errorPattern.MatchString(line) {
			a.Errors = append(a.Errors, line)
		}

		// スロークエリ検出
		if matches := slowQueryPattern.FindStringSubmatch(line); matches != nil {
			duration, _ := strconv.ParseFloat(matches[1], 64)
			if duration > 1000 { // 1秒以上
				a.SlowQueries = append(a.SlowQueries, SlowQuery{
					Duration: duration,
					Line:     line,
				})
			}
		}

		// 接続パターン分析
		if connectionPattern.MatchString(line) {
			if userMatches := userPattern.FindStringSubmatch(line); userMatches != nil {
				a.Connections[userMatches[1]]++
			}
		}
	}

	return scanner.Err()
}

func (a *PostgreSQLLogAnalyzer) AnalyzeErrors() {
	fmt.Println(strings.Repeat("=", 80))
	fmt.Println("エラー分析")
	fmt.Println(strings.Repeat("=", 80))

	errorTypes := make(map[string]int)
	for _, error := range a.Errors {
		lowerError := strings.ToLower(error)
		if strings.Contains(lowerError, "deadlock") {
			errorTypes["Deadlock"]++
		} else if strings.Contains(lowerError, "connection") {
			errorTypes["Connection Error"]++
		} else if strings.Contains(lowerError, "syntax") {
			errorTypes["Syntax Error"]++
		} else if strings.Contains(lowerError, "permission") {
			errorTypes["Permission Denied"]++
		} else {
			errorTypes["Other"]++
		}
	}

	fmt.Printf("\n総エラー数: %d\n", len(a.Errors))
	fmt.Println("\nエラータイプ別集計:")
	for errType, count := range errorTypes {
		fmt.Printf("  %s: %d\n", errType, count)
	}

	if len(a.Errors) > 0 {
		fmt.Println("\n最新のエラー（最大5件）:")
		start := len(a.Errors) - 5
		if start < 0 {
			start = 0
		}
		for _, error := range a.Errors[start:] {
			if len(error) > 200 {
				error = error[:200]
			}
			fmt.Printf("  %s\n", error)
		}
	}
}

func (a *PostgreSQLLogAnalyzer) AnalyzeSlowQueries() {
	fmt.Println("\n" + strings.Repeat("=", 80))
	fmt.Println("スロークエリ分析")
	fmt.Println(strings.Repeat("=", 80))

	if len(a.SlowQueries) == 0 {
		fmt.Println("\nスロークエリは検出されませんでした。")
		return
	}

	fmt.Printf("\nスロークエリ数: %d\n", len(a.SlowQueries))

	// 実行時間でソート
	sort.Slice(a.SlowQueries, func(i, j int) bool {
		return a.SlowQueries[i].Duration > a.SlowQueries[j].Duration
	})

	fmt.Println("\n最も遅いクエリ（TOP 5）:")
	limit := 5
	if len(a.SlowQueries) < limit {
		limit = len(a.SlowQueries)
	}
	for i := 0; i < limit; i++ {
		query := a.SlowQueries[i]
		line := query.Line
		if len(line) > 300 {
			line = line[:300]
		}
		fmt.Printf("\n%d. 実行時間: %.2f ms\n", i+1, query.Duration)
		fmt.Printf("   %s\n", line)
	}
}

func (a *PostgreSQLLogAnalyzer) AnalyzeConnections() {
	fmt.Println("\n" + strings.Repeat("=", 80))
	fmt.Println("接続パターン分析")
	fmt.Println(strings.Repeat("=", 80))

	type userCount struct {
		User  string
		Count int
	}
	var users []userCount
	for user, count := range a.Connections {
		users = append(users, userCount{user, count})
	}
	sort.Slice(users, func(i, j int) bool {
		return users[i].Count > users[j].Count
	})

	fmt.Println("\nユーザー別接続数:")
	for _, uc := range users {
		fmt.Printf("  %s: %d\n", uc.User, uc.Count)
	}
}

func (a *PostgreSQLLogAnalyzer) GenerateReport() error {
	fmt.Println("\n" + strings.Repeat("=", 80))
	fmt.Println("PostgreSQL ログ分析レポート")
	fmt.Printf("分析ファイル: %s\n", a.LogFile)
	fmt.Printf("分析日時: %s\n", time.Now().Format("2006-01-02 15:04:05"))
	fmt.Println(strings.Repeat("=", 80))

	if err := a.ParseLog(); err != nil {
		return err
	}

	a.AnalyzeErrors()
	a.AnalyzeSlowQueries()
	a.AnalyzeConnections()

	// 推奨アクション
	fmt.Println("\n" + strings.Repeat("=", 80))
	fmt.Println("推奨アクション")
	fmt.Println(strings.Repeat("=", 80))

	if len(a.Errors) > 100 {
		fmt.Println("⚠️  エラーが多発しています。アプリケーションまたはDB設定を確認してください。")
	}

	if len(a.SlowQueries) > 50 {
		fmt.Println("⚠️  スロークエリが多数検出されました。インデックス最適化を検討してください。")
	}

	maxConnections := 0
	for _, count := range a.Connections {
		if count > maxConnections {
			maxConnections = count
		}
	}
	if maxConnections > 1000 {
		fmt.Println("⚠️  特定ユーザーの接続数が多い。コネクションプーリングを確認してください。")
	}

	return nil
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("使用方法: go run log_analyzer.go <log_file>")
		os.Exit(1)
	}

	logFile := os.Args[1]
	analyzer := NewPostgreSQLLogAnalyzer(logFile)

	if err := analyzer.GenerateReport(); err != nil {
		fmt.Fprintf(os.Stderr, "エラー: %v\n", err)
		os.Exit(1)
	}
}
